uniffi_bindgen/src/templates/ObjectTemplate.rs

```diff
@@ -8,6 +8,22 @@
 // error message when processing this generated code.
 {% let handle_map = format!("UNIFFI_HANDLE_MAP_{}", obj.name().to_uppercase()) -%}
 
+
+{% if !obj.threadsafe() %}
+// We want to mark this as `deprecated` - long story short, the only way to
+// sanely do this using `#[deprecated(..)]` is to generate a function with that
+// attribute, then generate call to that function in the object constructors.
+#[deprecated(
+    since = "0.8.0",
+    note = "Non thread-safe objects are deprecated - you should upgrade \
+            `{{ obj.name() }}` to be threadsafe, then add the `[ThreadSafe]` \
+            annotation. \
+            See https://github.com/mozilla/uniffi-rs/#thread-safety for more details"
+)]
+#[allow(non_snake_case)]
+fn note_nonthreadsafe_deprecation_{{ obj.name() }}() {}
```

Comment:
nit: please prefix this with `uniffi_` or similar; I don't think we're 100% in the habit of doing this for our internal-use symbols, but it seems worthwhile to try to converge towards a common pattern there when we can.